lib/hashserv/client.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (C) 2019 Garmin Ltd.
+#
+# SPDX-License-Identifier: GPL-2.0-only
+#
+
+import logging
+import socket
+import bb.asyncrpc
+from . import create_async_client
+
+
+logger = logging.getLogger("hashserv.client")
+
+
+class AsyncClient(bb.asyncrpc.AsyncClient):
+    MODE_NORMAL = 0
+    MODE_GET_STREAM = 1
+
+    def __init__(self):
+        super().__init__('OEHASHEQUIV', '1.1', logger)
+        self.mode = self.MODE_NORMAL
+
+    async def setup_connection(self):
+        await super().setup_connection()
+        cur_mode = self.mode
+        self.mode = self.MODE_NORMAL
+        await self._set_mode(cur_mode)
+
+    async def send_stream(self, msg):
+        async def proc():
+            self.writer.write(("%s\n" % msg).encode("utf-8"))
+            await self.writer.drain()
+            l = await self.reader.readline()
+            if not l:
+                raise ConnectionError("Connection closed")
+            return l.decode("utf-8").rstrip()
+
+        return await self._send_wrapper(proc)
+
+    async def _set_mode(self, new_mode):
+        if new_mode == self.MODE_NORMAL and self.mode == self.MODE_GET_STREAM:
+            r = await self.send_stream("END")
+            if r != "ok":
+                raise ConnectionError("Bad response from server %r" % r)
+        elif new_mode == self.MODE_GET_STREAM and self.mode == self.MODE_NORMAL:
+            r = await self.send_message({"get-stream": None})
+            if r != "ok":
+                raise ConnectionError("Bad response from server %r" % r)
+        elif new_mode != self.mode:
+            raise Exception(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using `Exception` and `BaseException` can make your code prone to errors and difficult to maintain. Instead, we recommend using one of the [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html) or creating a custom exception class that is derived from `Exception` or one of its subclasses.